flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -525,61 +525,99 @@ class ExprLowering {
   }
 
   template <Fortran::common::TypeCategory TC, int KIND>
-  mlir::Value
-  genval(const Fortran::evaluate::Constant<Fortran::evaluate::Type<TC, KIND>>
-             &con) {
-    // TODO:
-    // - array constant not handled
-    // - derived type constant
+  mlir::Value genScalar(
+      const Fortran::evaluate::Scalar<Fortran::evaluate::Type<TC, KIND>>
+          &value) {
```

Comment:
The name and type suggest this this is general purpose code to lower scalar typed expressions. It isn't (I don't think). It's just lowering scalar constant values.
I think this should take the Constant<T> and do the GetScalarValue() and has_value() as common code at the top.